csrc/graph/task_graph.cpp

```diff
@@ -0,0 +1,714 @@
+          "Ran out of ready tasks before completing ordering");
+
+      TaskGraph::TaskId next_task_id = -1;
+      for (const TaskGraph::TaskId ready_id : ready_tasks_) {
+        if (ready_id > backtracked_task_id) {
+          next_task_id = ready_id;
+          break;
+        }
+      }
+
+      // Reset backtracked_task_id
+      backtracked_task_id = -1;
+
+      if (next_task_id == -1) {
+        // There are no ready tasks with ID above the backtracked_task_id. This
+        // means it is time to backtrack
+
+        if (steps_.empty()) {
+          // If there is nowhere to backtrack it means we are done with the
+          // search
+          result_.exhaustive = true;
+          break;
+        }
+        backtracked_task_id = backtrack();
+        continue;
+      }
+
+      advance(next_task_id);
+
+      // If our high water mark is above best_usage, terminate early and
+      // backtrack
+      TaskGraph::Size hwm = steps_.back().high_water_mark;
+      TaskGraph::Size best_hwm = best_steps.empty()
+          ? std::numeric_limits<TaskGraph::Size>::max()
+          : best_steps.back().high_water_mark;
+
+      if (hwm > best_hwm) {
+        backtracked_task_id = backtrack();
+        continue;
+      }
+
+      // Our usage is at or below best. Have we completed an ordering? If
+      // so, update best_steps
+      if (steps_.size() == (size_t)graph_.numTasks() && hwm < best_hwm) {
+        best_steps = steps_;
+      }
+    }
+
+    // Record our best found steps
+    result_.steps = best_steps;
+
+    if (isDebugDumpEnabled(DebugDumpOption::TaskGraph)) {
+      Clock::time_point stop = Clock::now();
+      debug() << "Found these steps in "
+              << (std::chrono::duration_cast<std::chrono::milliseconds>(
+                      stop - start)
+                      .count())
+              << " ms:\n";
+      for (const TaskGraph::Step& step : result_.steps) {
+        debug() << "  " << step << "\n";
+      }
+      debug() << "The search contained " << result_.iterations
+              << " iterations and was ";
+      if (!result_.exhaustive) {
+        debug() << "NOT ";
+      }
+      debug() << "exhaustive" << std::endl;
+    }
+
+    // Validate final result
+    NVF_ERROR(result_.steps.size() == (size_t)graph_.numTasks());
+    validate();
+  }
+
+ private:
+  const TaskGraph& orig_graph_;
+  const TaskGraph graph_;
+  const bool debug_;
+  const bool validate_;
+  const int64_t max_time_us_;
+
+  TaskGraph::SortResult result_;
+  std::vector<TaskGraph::Step> steps_;
+
+  //! There is one entry here for each task and indicating how many
+  //! dependencies are currently unmet. When this reaches zero the task becomes
+  //! ready.
+  std::vector<TaskGraph::DataId> outstanding_dependencies_;
+
+  //! There is one entry here for each Data and indicating how many uses there
+  //! are remaining. When it reaches zero, the Data can be freed if allowed.
+  std::vector<TaskGraph::TaskId> future_uses_;
```

Comment:
```suggestion
  std::vector<int64_t> future_uses_;
```
According to your comment, each element is a count not a task ID.